File: ios/Classes/Features/SessionFrames.swift
import ADEUMInstrumentation
import Flutter

extension SwiftAppDynamicsAgentPlugin {
  func startSessionFrame(result: @escaping FlutterResult, arguments: Any?) {
    guard let properties = arguments as? Dictionary<String, Any> else {
      return
    }
    
    guard let name = properties["name"] as? String else {
      let error = FlutterError(code: "500", message: "Agent startSessionFrame() failed.", details: "Please provide a valid session name.")
      result(error)
      return
    }
    
    guard let id = properties["id"] as? String else {
      let error = FlutterError(code: "500", message: "Agent startSessionFrame() failed.", details: "Please provide a valid session ID.")
      result(error)
      return
    }
    
    let sessionFrame: ADEumSessionFrame = ADEumInstrumentation.startSessionFrame(name)
    sessionFrames[id] = sessionFrame
    result(nil)
  }
  
  func updateSessionFrameName(result: @escaping FlutterResult, arguments: Any?) {
    guard let properties = arguments as? Dictionary<String, Any> else {
      return
    }
    
    guard let newName = properties["newName"] as? String else {
      let error = FlutterError(code: "500", message: "Agent updateSessionFrameName() failed.", details: "Please provide a valid session name.")
      result(error)
      return
    }
    
    guard let id = properties["id"] as? String else {
      let error = FlutterError(code: "500", message: "Agent updateSessionFrameName() failed.", details: "Please provide a valid session ID.")
      result(error)
      return
    }
    
    sessionFrames[id]?.updateName(newName)
    result(nil)
  }
  
  func endSessionFrame(result: @escaping FlutterResult, arguments: Any?) {
    guard let id = arguments as? String else {
      let error = FlutterError(code: "500", message: "Agent endSessionFrame() failed.", details: "Please provide a valid session ID.")
      result(error)
      return
    }
    
    if let sessionFrame = sessionFrames[id] {
      sessionFrame.end()
      sessionFrames[id] = nil
    }
    result(nil)
  }
}

File: android/src/main/kotlin/com/appdynamics/appdynamics_agent/features/SessionFrames.kt
/*
 * Copyright (c) 2022. AppDynamics LLC and its affiliates.
 * All rights reserved.
 *
 */

package com.appdynamics.appdynamics_agent.features

import androidx.annotation.NonNull
import com.appdynamics.appdynamics_agent.AppDynamicsAgentPlugin
import com.appdynamics.eumagent.runtime.Instrumentation
import io.flutter.plugin.common.MethodChannel

fun AppDynamicsAgentPlugin.startSessionFrame(
    @NonNull result: MethodChannel.Result,
    arguments: Any?
) {
    val properties = arguments as HashMap<*, *>
    val name = properties["name"] as? String
    val id = properties["id"] as? String

    if (name == null) {
        result.error(
            "500",
            "Agent startSessionFrame() failed.",
            "Please provide a valid session name."
        )
        return
    }

    if (id == null) {
        result.error(
            "500",
            "Agent startSessionFrame() failed.",
            "Please provide a valid session ID."
        )
        return
    }

    val sessionFrame = Instrumentation.startSessionFrame(name)
    sessionFrames[id] = sessionFrame

    result.success(null)
}

fun AppDynamicsAgentPlugin.updateSessionFrameName(
    @NonNull result: MethodChannel.Result,
    arguments: Any?
) {
    val properties = arguments as HashMap<*, *>
    val newName = properties["newName"] as? String
    val id = properties["id"] as? String

    if (newName == null) {
        result.error(
            "500",
            "Agent updateSessionFrame() failed.",
            "Please provide a valid session name."
        )
        return
    }

    if (id == null) {
        result.error(
            "500",
            "Agent updateSessionFrame() failed.",
            "Please provide a valid session ID."
        )
        return
    }

    sessionFrames[id]?.updateName(newName)
    result.success(null)
}

fun AppDynamicsAgentPlugin.endSessionFrame(
    @NonNull result: MethodChannel.Result,
    arguments: Any?
) {
    val id = arguments as? String

    if (id == null) {
        result.error(
            "500",
            "Agent endSessionFrame() failed.",
            "Please provide a valid session ID."
        )
        return
    }

    if (sessionFrames.containsKey(id)) {
        sessionFrames[id]?.end()
        sessionFrames.remove(id)
    }
    result.success(null)
}